src/execution_algorithms/kfac/kfac_block_fc_conv.cpp

```diff
@@ -369,6 +379,293 @@ void kfac_block_fc_conv<Device>::update_kronecker_inverse(
+      for (auto& input : this->m_parent_local_activations) {
+        if(dtl_parent.get_data_layout() == data_layout::DATA_PARALLEL){
+          input = make_unique<El::DistMatrix<DataType, El::STAR, El::VC, El::ELEMENT, Device>>(comm->get_secondary_grid(), 0);
+          if(comm->enable_subgrid_async_communication())
+            this->m_subset_matrix.push_back(
+                make_unique<El::DistMatrix<DataType, El::STAR, El::VC, El::ELEMENT, Device>>(comm->get_subset_grid(), 0));
+        }
+        else{
+          input = make_unique<El::DistMatrix<DataType, El::MC  , El::MR, El::ELEMENT, Device>>(comm->get_secondary_grid(), 0);
+          if(comm->enable_subgrid_async_communication())
+            LBANN_ERROR("Async prgoress is not supported for model-parallel layer layout in sub-grid parallelism");
+        }
+      }
+    }
+
+    std::vector<El::mpi::Request<DataType>> requests;
+
+    if(comm->enable_subgrid_async_communication()==false)
+    {
+      El::Copy(local_activations,*this->m_parent_local_activations[0]);
+    }
+    else{
+      const auto local_activations_vc = dynamic_cast<const El::DistMatrix<DataType, El::STAR, El::VC, El::ELEMENT, Device>*>(&(local_activations));
+      auto local_activations0 = dynamic_cast<El::DistMatrix<DataType, El::STAR, El::VC, El::ELEMENT, Device>*>(&(*this->m_parent_local_activations[0]));
+      auto subset0 = dynamic_cast<El::DistMatrix<DataType, El::STAR, El::VC, El::ELEMENT, Device>*>(&(*this->m_subset_matrix[0]));
+
+      kfac::TranslateBetweenGridsVCAsync(*local_activations_vc,
+                                              *local_activations0,
+                                              *subset0,
+                                              this->m_requests_forward_end);
+
+      
+    }//Async progress
+  }
+
+  if(comm->get_grid_type() == GridType::NO_GRID or comm->get_KFAC_subgrid_create_two_models()){
+
+    for (auto& input : this->m_parent_local_activations) {
+      if(dtl_parent.get_data_layout() == data_layout::DATA_PARALLEL)
+        input = make_unique<El::DistMatrix<DataType, El::STAR, El::VC, El::ELEMENT, Device>>(comm->get_trainer_grid(), 0);
+      else
+        input = make_unique<El::DistMatrix<DataType, El::MC  , El::MR, El::ELEMENT, Device>>(comm->get_trainer_grid(), 0);
+    }
+    El::LockedView( *(this->m_parent_local_activations[0]), local_activations );
+  } 
+  
+}
+
+template <El::Device Device>
+void kfac_block_fc_conv<Device>::end_communication_forward_end(
+    lbann_comm* comm) {
+  if((comm->get_grid_type() == GridType::SECONDARY_GRID or comm->get_grid_type() == GridType::PRIMARY_GRID) 
+      and comm->enable_subgrid_async_communication()
+      and comm->get_KFAC_subgrid_create_two_models()==false){
+    auto primary_grid_ranks = comm->get_primary_grid_ranks();
+    auto secondary_grid_ranks = comm->get_secondary_grid_ranks();
+
+    for(auto& req:this->m_requests_forward_end){
+      ::Al::Wait<::Al::NCCLBackend>(req);
```

Comment:
I see build errors on Catalyst since `Al::NCCLBackend` isn't defined. They should all be replaced with `BackendT`:
```suggestion
      ::Al::Wait<BackendT>(req);
```